src/sentry/rules/processing/buffer_processing.py

```diff
@@ -141,8 +140,77 @@ def process_in_batches(project_id: int, processing_type: str) -> None:
             )
 
 
+def to_log_format(project_ids: list[tuple[int, float]]) -> str:
+    return ", ".join(f"{project_id}: {timestamp}" for project_id, timestamp in project_ids)
+
+
+def clear_processed_project_ids(
+    buf: Buffer, processing_type: str, processed: list[tuple[int, float]], buffer_key: str
+) -> None:
+    """
+    Safely clear processed project IDs from the buffer.
+    We want to remove all of the projects we've processed unless they've been updated since.
+    This can typically be done by deleting all up to the latest processed timestamp, but it's
+    possible that a new update has been added within that time window since we checked, so this
+    method errs toward only removing projects in a range we know doesn't include any new projects.
+
+    This may rarely result in duplicate scheduling, but that is expected to be safe.
+    """
+    if not processed:
+        return
+    max_processed_ts = max(timestamp for _, timestamp in processed)
+    expected_deleted = len(processed)
+    # Delete all project IDs that we've processed unless they've been updated since.
+    # Any newly added projects should be above max_observed and will be processed on
+    # next execution.
+    to_delete = buf.get_sorted_set(buffer_key, max=max_processed_ts)
+    new_project_ids = {proj_id for proj_id, _ in to_delete} - {proj_id for proj_id, _ in processed}
+    if new_project_ids:
+        # We had some older timestamps show up after our fetch.
+        logger.warning(
+            "buffer_processing.project_id_list_new_project_ids",
+            extra={
+                "project_ids": to_log_format(processed),
+                "new_project_ids": new_project_ids,
+                "processing_type": processing_type,
+            },
+        )
+        max_safe_ts = None
+        min_new_proj_ts = min(
+            timestamp for proj_id, timestamp in to_delete if proj_id in new_project_ids
+        )
+        for _, timestamp in to_delete:
+            if timestamp >= min_new_proj_ts:
+                break
+            max_safe_ts = max(max_safe_ts or timestamp, timestamp)
+        if max_safe_ts is None:
+            # Something went wrong.
+            logger.error(
+                "Can't safely clear projects from buffer",
+                extra={
+                    "project_ids": to_log_format(processed),
+                    "new_project_ids": new_project_ids,
+                    "processing_type": processing_type,
+                },
+            )
+            return
```

Comment:
Since this affects the existing delayed processing, should the default behavior be to clear up from when we last processed? This is what we were doing before and I'm worried if we hit this we won't ever clear projects from the buffer